In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import collections

In [2]:
dataset = pd.read_csv('Social_Network_Ads.csv')
X = dataset.iloc[:, 2:4].values.astype(np.float64)
y = dataset.iloc[:, 4].values.astype(np.float64)
dataset

User ID  Gender  Age  EstimatedSalary  Purchased
0    15624510    Male   19            19000          0
1    15810944    Male   35            20000          0
2    15668575  Female   26            43000          0
3    15603246  Female   27            57000          0
4    15804002    Male   19            76000          0
5    15728773    Male   27            58000          0
6    15598044  Female   27            84000          0
7    15694829  Female   32           150000          1
8    15600575    Male   25            33000          0
9    15727311  Female   35            65000          0
10   15570769  Female   26            80000          0
11   15606274  Female   26            52000          0
12   15746139    Male   20            86000          0
13   15704987    Male   32            18000          0
14   15628972    Male   18            82000          0
15   15697686    Male   29            80000          0
16   15733883    Male   47            25000          1
17   15617482    Male   45            26000          1
18   15704583    Male   46            28000          1
19   15621083  Female   48            29000          1
20   15649487    Male   45            22000          1
21   15736760  Female   47            49000          1
22   15714658    Male   48            41000          1
23   15599081  Female   45            22000          1
24   15705113    Male   46            23000          1
25   15631159    Male   47            20000          1
26   15792818    Male   49            28000          1
27   15633531  Female   47            30000          1
28   15744529    Male   29            43000          0
29   15669656    Male   31            18000          0
..        ...     ...  ...              ...        ...
370  15611430  Female   60            46000          1
371  15774744    Male   60            83000          1
372  15629885  Female   39            73000          0
373  15708791    Male   59           130000          1
374  15793890  Female   37            80000          0
375  15646091  Female   46            32000          1
376  15596984  Female   46            74000          0
377  15800215  Female   42            53000          0
378  15577806    Male   41            87000          1
379  15749381  Female   58            23000          1
380  15683758    Male   42            64000          0
381  15670615    Male   48            33000          1
382  15715622  Female   44           139000          1
383  15707634    Male   49            28000          1
384  15806901  Female   57            33000          1
385  15775335    Male   56            60000          1
386  15724150  Female   49            39000          1
387  15627220    Male   39            71000          0
388  15672330    Male   47            34000          1
389  15668521  Female   48            35000          1
390  15807837    Male   48            33000          1
391  15592570    Male   47            23000          1
392  15748589  Female   45            45000          1
393  15635893    Male   60            42000          1
394  15757632  Female   39            59000          0
395  15691863  Female   46            41000          1
396  15706071    Male   51            23000          1
397  15654296  Female   50            20000          1
398  15755018    Male   36            33000          0
399  15594041  Female   49            36000          1

[400 rows x 5 columns]

In [3]:
# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test =  train_test_split(X,y,test_size = 0.25, random_state = 54)

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

In [4]:
def classifier(X_train , y_train, X ,r):
    def func(x,y):
        return (x - X[0])**2 + (y-X[1])**2 - r**2
    L=[]
    for i in range(len(X_train[:,0])):
        x = X_train[i,0]
        y = X_train[i,1]
        
        if func(x,y)<=0:
            L.append([x,y,y_train[i]])
    y1 = 0
    y0=0        
    for i in L:
        if i[2] == 1:
            y1+=1
        else:
            y0+=1
    
    Px = len(L)/len(X_train[:,0])
    P1 =  collections.Counter(y_train)[1]/ len(y_train)
    P0 =  collections.Counter(y_train)[0]/ len(y_train)    
    Px1 = y1 / collections.Counter(y_train)[1]
    Px0 = y0 / collections.Counter(y_train)[0]
    
    p1x = Px1*P1/Px
    p0x = Px0*P0/Px
    if p1x>=p0x:
        return 1
    else:
        return 0

def predict(X_train,y_train,X_test,r):
    y_predict = np.ones(len(X_test[:,0]))
    for i in range(len(X_test[:,0])):
        X = X_test[i,:]
        if classifier(X_train, y_train , X,r)==0:
            y_predict[i] = 0
    return y_predict

In [5]:
y_pred = predict(X_train,y_train,X_test,0.5)
pd.DataFrame(y_pred)

0
0   1.0
1   0.0
2   0.0
3   0.0
4   0.0
5   0.0
6   0.0
7   0.0
8   1.0
9   1.0
10  0.0
11  0.0
12  0.0
13  0.0
14  0.0
15  1.0
16  1.0
17  1.0
18  1.0
19  1.0
20  1.0
21  0.0
22  0.0
23  0.0
24  0.0
25  0.0
26  0.0
27  1.0
28  0.0
29  0.0
..  ...
70  1.0
71  0.0
72  0.0
73  1.0
74  0.0
75  0.0
76  0.0
77  0.0
78  0.0
79  0.0
80  0.0
81  1.0
82  1.0
83  0.0
84  1.0
85  1.0
86  1.0
87  0.0
88  0.0
89  0.0
90  1.0
91  1.0
92  0.0
93  0.0
94  0.0
95  0.0
96  0.0
97  0.0
98  1.0
99  0.0

[100 rows x 1 columns]

In [6]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
cm

array([[60,  4],
       [ 3, 33]])

No of correct observations = Sum of left Diagonal = 93

No of incorrect predictions = Sum of Right Diagonal = 7

Accuracy = 93%